### ISSUE TITLE
Consider changing `Char` to represent a code point rather than a UTF-16 code unit

### ISSUE THREAD

**Issue by natefaubion** [purescript/purescript#3662]
Reference: https://github.com/pure-c/purec/issues/34#issuecomment-497961729

Exposing `Char` as UTF-16 code units is problematic for backends that need to use something _other_ than UTF-16 as the underlying representation. Would it be better if `Char` represented was encoding agnostic and represented a code point? What would that mean for `purescript-strings`?

**Comment by anttih** [comment]
For `purescript-strings` could we do the following?

* In `CodePoints`, get rid of `CodePoint` and use `Char`. Implement all functions without referring to `CodeUnits` module.
* Move `CodeUnits` module to a separate library. There we could have a `CodeUnit` newtype around `Int` and different backends could implement the code unit API how ever they wanted.

**Comment by andyarvanitis** [comment]
@anttih 
> For `purescript-strings` could we do the following?
> 
> * In `CodePoints`, get rid of `CodePoint` and use `Char`...

I'd say keeping
```Haskell
newtype CodePoint = CodePoint Int
```
is ok, as an integer codepoint representation is universal to all unicode encodings.

And it's best not to assume that a `Char` is literally a 32-bit integer. For example, right now in the C++ backend `Char` maps to a (single) *UTF-8* "entity", e.g. `u8"Å"` (it was a `char32_t` in pure11). This could change, but maybe let's keep `Char` an encoding-agnostic code point, as @natefaubion brings up.

**Comment by andyarvanitis** [comment]
Also, code units are encoding-specific – I don't know if you can really abstract those. So if you have separate libs for those, they should be for UTF-32, UTF-16, UTF-8, it seems...

**Comment by anttih** [comment]
> And it's best not to assume that a Char is literally a 32-bit integer.

But shouldn't almost all code units fit into a 32-bit integer, be it UTF-8, 16 or 32? That's why I thought `Int` would be enough for a code unit. But yeah, I guess it's not optimal if your code unit is smaller than 32 bits. Maybe it's up to the different backends to provide their own code unit libraries then and have purescript deal with code points all the time.

> This could change, but maybe let's keep Char an encoding-agnostic code point, as @natefaubion brings up.

That's what I'm saying, the built-in type `Char` would be a code point. Why have a separate `CodePoint` type if `Char` is a code point?

**Comment by michaelficarra** [comment]
What does this mean for JavaScript interop? How will JavaScript strings be represented in PureScript? It's really nice right now having a zero-overhead FFI to JavaScript APIs that produce JavaScript strings.

**Comment by andyarvanitis** [comment]
@anttih I really don't have an opinion about `newtype CodePoint`, I just meant that it shouldn't need to change (there might be people who want/need it, I don't actually know). Also, translations to/from `Int` and code points – for whatever encoding – are always the same.

As for `Char`, my only point was that it shouldn't be assumed to be an integer of some sort "under-the-hood." For both the javascript and the C++ backends, it's currently just a string containing one "character."

**Comment by anttih** [comment]
@michaelficarra 

> What does this mean for JavaScript interop? How will JavaScript strings be represented in PureScript? It's really nice right now having a zero-overhead FFI to JavaScript APIs that produce JavaScript strings.

This shouldn't change at all. I think the point would just be to abstract `Char` to be a code point instead of it being 16-bit code unit. In JS you'd still be dealing with UTF-16 strings, so no overhead there, but in other backends you could more easily use some other encoding like UTF-8.

**Comment by michaelficarra** [comment]
@anttih I am talking about consuming JS strings. If PureScript strings are sequences of PureScript chars, and PureScript chars are Unicode code points, PureScript strings cannot be used to represent JavaScript strings with unpaired surrogates. Unless we'd represent them with the reserved surrogate code point? But then we'd still have to be aware of the concatenation rules of the underlying representation.

**Comment by timjs** [comment]
The general trend of languages created in the last decade, is to take `Char` to represent a [*Code Point*](http://www.unicode.org/glossary/#code_point) or [*Unicode Scalar Value*](http://www.unicode.org/glossary/#unicode_scalar_value), stored as a 32-bit (unsigned) integer. See for example:

* Swift's [`Character`](https://developer.apple.com/documentation/swift/character)
* Rust's [`char`](https://doc.rust-lang.org/std/primitive.char.html)
* Golang's [`rune`](https://golang.org/pkg/builtin/#rune)
* Crystal's [`Char`](https://crystal-lang.org/api/0.29.0/Char.html)

A string then, is taken to be a collection over characters. So iterating over strings yields Code Points. How strings are internally encoded, is an implementation detail (which leaks more through, depending on the language):

* Swfit's [`String`](https://developer.apple.com/documentation/swift/string) used to be ASCII or UTF-16 encoded with multiple "views" on them ([`UnicodeScalarView`](https://swiftdoc.org/v3.1/type/string.unicodescalarview/), [`UTF16View`](https://swiftdoc.org/v3.1/type/string.utf16view/), [`UTF8View`](https://swiftdoc.org/v3.1/type/string.utf8view/)) [but are all UTF-8 now](https://swift.org/blog/utf8-string/)
* Rust's [`str`](https://doc.rust-lang.org/std/primitive.str.html), Golang's [`string`](https://golang.org/pkg/builtin/#string), and Crystal's [`String`](https://crystal-lang.org/api/0.29.0/String.html) are all UTF-8 encoded byte arrays (although it [is possible](https://blog.golang.org/strings) to convert [non-valid UTF-8](https://crystal-lang.org/api/0.29.0/String.html) byte arrays into strings).

I'd imagine that PureScript provides a `Char` type representing a Code Point, and an *opaque* `String` type with operations using `Char`s (Code Points). Backends could than choose how `String`s should be encoded: UTF-16 for JavaScript, UTF-8 for C/C++, etc. Each backend could provide a *fast* interface by providing direct access to it's internal representation: `CodeUnit` for JavaScript, `Byte` for C/C++. These would not work across backends.

Alternatively, we should choose one representation and force that onto every backend, which would impose less freedom on backend developers.

**Comment by anttih** [comment]
@michaelficarra Ok, I see your point, and it's a good one. No idea how to solve that one. We would need to somehow abstract the splitting of a string to code points so that the JS backend could deal with lone surrogates and the other backends could just deal without them.

**Comment by andyarvanitis** [comment]
I'd recommend we keep `Char` an opaque type too (like `String`), with it representing a unicode code point – don't assume UTF-32, it would be up to backends as to how they actually want to encode it.  Conversions to/from `Int` would be what you would expect (a 32-bit unicode value. which *is* effectively a single UTF-32 entity).

Also keep in mind that javascript doesn't even have a char type, just strings – although it does have functions that operate on "chars" (UTF-16 code units) – but I'd involve javascript experts like @michaelficarra in how best to break things out to keep js interop good.

Oh, and I think abstracting code ***units*** is kind of a bad idea, in general. It's dependent on your encoding, so I think there should be separate appropriate libs for UTF-16, UTF-8, etc., and they don't even need to be backend-specific.

Please keep in mind this is just my two cents.

**Comment by hdgarrood** [comment]
> I am talking about consuming JS strings. If PureScript strings are sequences of PureScript chars, and PureScript chars are Unicode code points, PureScript strings cannot be used to represent JavaScript strings with unpaired surrogates. Unless we'd represent them with the reserved surrogate code point? But then we'd still have to be aware of the concatenation rules of the underlying representation.

Yes, I'm imagining that we would represent them with the reserved surrogate code point. This behaviour would be specific to the JS backend and not part of the language definition; I think the language spec (if we ever get around to actually writing one) should probably say that the behaviour of strings which contain reserved surrogate code points is unspecified, although the JS backend would continue to behave in the way it does now.

**Comment by hdgarrood** [comment]
I think there is a problem we'd need to solve first, which is that there are currently insufficient tools for efficiently iterating over strings in an backend- and encoding-agnostic way. Currently, with the API provided by Data.String.CodePoints, almost anything you might want to do incurs excessive copying; either you convert the entire string to an array of code points, or you use functions like `uncons` and `dropWhile` and so on, which copy the tail of the string; neither option is very appealing if you have a long string and you don't necessarily need to look at the entire thing. I think that at the moment, if you want to write efficient String code, you basically have to work with the underlying bytes. The `string-parsers` library is a good example: right now you have to choose between good performance (`Text.Parsing.StringParser.CodeUnits`) and not accidentally splitting up surrogate pairs (`Text.Parsing.StringParser.CodePoints`).

**Comment by hdgarrood** [comment]
> or you use functions like uncons and dropWhile and so on, which copy the tail of the string;

In JS, strings are immutable, so functions like `uncons` and `dropWhile` are likely to be constant-time after all. See https://github.com/purescript/purescript-strings/issues/120

**Comment by drathier** [comment]
What would it take to make this change, or decide whether to go through with this change?

Questions:
- How do we deal with unpaired surrogates? Do people need to handle them in practice?

Suggested change:
- Change Char to be an opaque code point instead of a utf-16 code unit, actual representation being backend-defined.

Purerl already uses utf-8 strings, and seemingly the C++ backend too, so we're already torn apart on this.

**Comment by natefaubion** [comment]
Would another approach be to add CodePoint literal support?

**Comment by drathier** [comment]
That could work, but I'm wondering if we'd run into the same issues as we do today; we'd have to make sure all code converting between `CodeUnit`s and `CodePoint`s are in FFI, which isn't the case today. 

I'm also wondering what the syntax would be like for multiple char-like literals. Which one gets the usual syntax, i.e. `'c'`? Do they share syntax, similar to `OverloadedStrings` in ghc?

**Comment by natefaubion** [comment]
My main reason for making the suggestion is to reduce inertia. I'd say this is a pain that's felt by a comparatively small number of PureScript users, at the moment, which is why there's so little movement on this issue. The effort vs impact of the original proposal isn't skewed favorably, and as far as I know, no one is really championing this or volunteering to see it through.

> we'd have to make sure all code converting between CodeUnits and CodePoints are in FFI, which isn't the case today.

Is this a pain that's felt by users, or a pain that's felt by backend implementations? It's not clear to me why this is inherently a con.

> I'm also wondering what the syntax would be like for multiple char-like literals. Which one gets the usual syntax, i.e. 'c'? Do they share syntax, similar to OverloadedStrings in ghc?

Like Int vs Number, we would have some sort of syntax to distinguish them. I've not made a suggestion for that because it's bikeshed city.

**Comment by chtenb** [comment]
It seems to me that many of the inconveniences arise from the fact that we have one string type with multiple views on it. Some programming languages, like Python, solve this by having two different string types. Python has a `String` and a `ByteString` type. The `String` type represents a unicode string. The encoding that is used internally is not exposed to the outside world and is guaranteed to be valid because all exposed operations on the `String` type are on the CodePoint level. So you don't have to deal with lone surrogates and other malformations once you have obtained a value of type `String`.

`String` also exposes a `encode(encoding='utf8')` function, which allows you to convert the string into a `ByteString`. This is more like an array of bytes and it exposes operations on the byte level. In return, the `ByteString` type exposes `decode(encoding='utf8')`, which may fail if the input is not valid encoded, but returns a `String` value in case of success.

There are different literal notations of each type. `'this is a String'` and `b'this is a bytestring'`. Python also doesn't really have a `Char` type, they only have a string of length 1.

I'm not sure how much of this maps well onto the PureScript case, but the idea of more than one string type doesn't seem to far fetched for a strongly typed programming language. I would imagine that the internal representation used by the `String` type would differ per PureScript backend (but this wouldn't be noticable since it doesn't leak). So if some foreign JS function hands us a string, we would treat it like a `ByteString` and call `decode UTF16` on it to obtain a safe `String`. In the JS case this decode function wouldn't have to do anything apart from checking that the encoding is valid, since the internal representation was already UTF16.

Again, I'm not sure if this model is a good fit for PureScript, but I wanted to throw it out here since it seems a relevant consideration.

**Comment by michaelficarra** [comment]
@chtenb That's an elegant solution that sounds like it would actually work pretty well for us. We might've gotten a bit too tunnel visioned on the goal of zero overhead FFI when we designed the current model.

**Comment by garyb** [comment]
Switching to having different types does make a lot of sense to me.

**Comment by drathier** [comment]
This is a much harder problem than it first seemed to me. I'll start with some base assertions:
- The main problem atm is that different backends use different encodings, and thus any code that uses Chars is backend-specific, but it's written in purescript and not in the local ffi language, so portability really suffers.
- Free string ffi would still be nice, so let's try to keep that.
- String slicing and indexing might be slow on code points based strings; implementing indexing and slicing in ffi might save some performance on utf32 backends, but probably not for js/utf16, so I'd suggest not having slicing or indexing functions on code point based strings.
- It's fine for ffi code to assume string encoding; it should always be the native encoding per backend. This is how we get free string ffi today.

Re encoding-specific strings (StringUtf8, StringUtf16, etc):
- This gives back slicing and indexing. This means we'll might get hand-optimized code that's faster on a particular backend but slower on others. Using StringUtf8 with the Erlang backend will be fast, but running the same StringUtf8-based code on the JS backend would probably be very slow. However, unlike today, it would still be correct, even if it's slow. Correctness matters much more than performance. The more likely scenario of JS code using StringUtf16 for performance reasons would be much less of a hit to utf8-based backends though than the other way around.

Re bytestrings:
- Similarly to StringUtf8/CharUtf8, we can add ByteString/Byte. I see no issues here except maybe what the syntax should be, but let's save that for another day.

Summary and proposed approach:
- `String`s are native encoding, they expose code points only, via `Char`
- `String`s passed either way through ffi uses native encoding
- If you want non-native encoding or to be specific about which encoding you want, you use e.g. `StringUtf8`/`CharUtf8`; these types are also allowed in ffi and the conversion to/from native string types happens in ffi code if needed; I'm sure some backends will not convert to the native string type and back in these cases, e.g. c++.
- We add conversion functions between e.g. `StringUtf8` and `String` in ffi code, and they're identity functions for e.g. `String.toUtf16`/`String.fromUtf16` on js/utf16 backends.
- Since `String` and `Char` are the shorter names, they'll be the default that people use.

Further questions:
- Does this interact well with something like Haskell's `IsString`, if we added that in the future? All string literals default to `String` like in ghc unless type inference figures out that it should be another string type?
- Are code points the right abstraction? Should we go all the way to graphemes? Slicing an [emoji family](https://emojipedia.org/families/) in half when adding an ellipsis (…) to shorten a string might be quite sad, and whichever one we call `String` is for sure the one people will use the most. Maybe a grapheme-based string api can be implemented no-problem on top of a code point based string, reusing the same type? There's no encoding issue there at least.

**Comment by chtenb** [comment]
> String slicing and indexing might be slow on code points based strings; implementing indexing and slicing in ffi might save some performance on utf32 backends, but probably not for js/utf16, so I'd suggest not having slicing or indexing functions on code point based strings.

If the backend already exposes codepoint-level operations on the native string, why wouldn't we want to expose those through `String`? E.g. JS has a [`codePointAt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt) function.

> Re encoding-specific strings (StringUtf8, StringUtf16, etc)

What would be the benefit of having string types that deviate from the native encoding? The only thing I can think of is when you are handed a binary string from somewhere in some non-native encoding you wouldn't have to convert the encoding but could still use codepoint-level operations. Are there others? It does sound like a niche performance optimization, but I might be wrong.

> The more likely scenario of JS code using StringUtf16 for performance reasons

In your proposal the `String` implementation would be a native string, so utf16 in the JS case. So what performance reasons would there be left to use the explicit `StringUtf16` type on the JS backend?

> We add conversion functions between e.g. StringUtf8 and String in ffi code, and they're identity functions for e.g. String.toUtf16/String.fromUtf16 on js/utf16 backends.

I'm inclined to agree, but I'm not sure if the correctness holds up a 100% of the time. I think JS strings are not guaranteed to not have lone surrogates. I think we maybe would want to forbid `String` to contain codepoints which represent a lone surrogate? (i.e. which are not valid unicode)

> Maybe a grapheme-based string api can be implemented no-problem on top of a code point based string, reusing the same type? There's no encoding issue there at least.

Yeah, that sounds like the most straightforward solution.

**Comment by MonoidMusician** [comment]
I don't think it's sufficient to only expose a code point API on strings, because a lot of FFI code will only return indexes appropriate for the native views. For example, the selection API for DOM input fields will return native UTF-16 indexes in JS, and it would be expensive/difficult to convert that back to code points.

Is there some way to design an API where there is still a single string type with two views on it: the default code point API (with no performance guarantees), and a native API that has an abstract `CharPart` (representing native indexing) with functions to aggregate `CharPart` into code points. I think this combination would allow doing the right thing in terms of code points as well as getting better performance through the native API, with the caveat that the latter must be used correctly so it doesn't get out of sync with the surrogates/etc.

I'm envisioning some kind of API that says “After a sensible boundary (e.g. the beginning of the string and after parsing full codepoints), we got a `CharPart` that is either a complete code point by itself, or expecting 1 or 2 or 3 more `CharPart`s to make a full code point”. Then a parser layered on top of this would be able to quickly scan using native indexing, using that new API to ensure it is respecting the code point boundaries of the native encoding, without having to know what the encoding is exactly.

In the larger picture, I would then advocate for removing `Char` literals in favor of code point literals, but otherwise keeping the API mostly the same. I named the native char `CharPart` above for clarity, but I would actually think it could just be `Char` if we drop the guarantee that `Char` should represent UTF-16 code units and instead leave it for whatever the native encoding chunk is. We just need a more comprehensive `Char[Part] -> CodePoint` API then. (Our currently conversion is already subtly broken haha.)

Do you think it would work for your needs?

**Comment by chtenb** [comment]
> Is there some way to design an API where there is still a single string type with two views on it:

Is it necessary to have two views on the same type? Perhaps we could have a `StringUtf16` type for this scenario, in the spirit of how @drathier described it. Conversion from `String` to `StringUtf16` is free in JS because the underlying encoding does not change, and by having different types you make the change in API explicit.

**Comment by MonoidMusician** [comment]
I didn't mention it, but I would like to avoid having specific encodings in PureScript. I don't see how it solves the problems we are facing, if it is (1) another thing that backends need to support or (2) something that the PS code would need to magically divine whether it is appropriate for the backend.

I think the primary tension is between fast native access and correct code point access, not between what encoding is used for the underlying string. And I think a single `String` type with two different APIs would satisfy that dichotomy and also closely resembles what we have today already. But if there are certain need-to-haves that my proposal doesn't address I'd like to hear about it.

**Comment by chtenb** [comment]
> But if there are certain need-to-haves that my proposal doesn't address I'd like to hear about it.

Perhaps not a hard need-to-have, but my personal strong would-like-to-have is a wellformedness guarantee on `String` (i.e. that the string has a valid encoding), because
- this would eliminate a class of bugs for free
- would make it more easy to write correct code unit level logic (see below)

If we expose functions like `appendCodeUnit :: String -> CodeUnit -> String` on our `String` type, I don't think we can hold this wellformedness guarantee.

However, as you point out, it would still be very useful to expose code unit level accessor functions like `codeUnitAt :: String -> Int -> CodeUnit`, for FFI and for certain performance optimizations. Since such functions are "readonly" we don't risk forming malformed strings and the wellformedness guarantee of the `String` type remains standing. Would such accessor functions be enough to cover the FFI usecases in your experience @MonoidMusician ? If we also need code unit level functions that produce strings, perhaps we could have a separate `CodeUnitString` type for that, such that we can have `codeUnitSlice :: Int -> Int -> String -> CodeUnitString` and `appendCodeUnit :: String -> CodeUnit -> CodeUnitString`.

The funny thing is that the wellformedness guarantee of `String` would make it easier to write correct logic with code unit level accessors too, because one does not have to account for malformations in the original string.

So to summarize, `String` would be a `CodeUnitString` with the additional guarantee that the string is wellformed. To me this seems like a natural thing to want to have in PureScript, just like we have `List` and `NonEmptyList`, where the latter is just the former with the additional guarantee that the list has at least one element.

**Comment by MonoidMusician** [comment]
Parsers need to remember what they've parsed (e.g. identifiers), so I'm not sure a read-only API would suffice. You basically would need to bundle the iterator pattern in the API to guarantee safety, otherwise how would you know that they are always advancing code unit indices by code point boundaries?

I think some amount of care is required around code-unit manipulation on the part of the programmer, no matter how the API looks, so I'm not convinced that 100% safety is something a core library could provide. Although I would be open to adding a wellformed predicate to double check whether a string is properly encoded, that each backend implements appropriately (or ignores).

Why not create a separate library with safe strings and a limited API as you suggest? I suppose you would like literal syntax – I'm not sure what guarantees the PS compiler currently provides around those?

**Comment by rhendric** [comment]
> I'm not sure what guarantees the PS compiler currently provides around those?

You can write isolated UTF-16 surrogates in strings by using "\xDCBA" escapes, and the compiler won't stop you. I don't see a compelling reason not to change that, though.

**Comment by michaelficarra** [comment]
> Although I would be open to adding a wellformed predicate to double check whether a string is properly encoded

There is a TC39 proposal to add this to the JS standard library as well, so it will probably be a constant time operation in that backend (eventually): https://github.com/guybedford/proposal-is-usv-string

**Comment by chtenb** [comment]
> Parsers need to remember what they've parsed (e.g. identifiers), so I'm not sure a read-only API would suffice. You basically would need to bundle the iterator pattern in the API to guarantee safety, otherwise how would you know that they are always advancing code unit indices by code point boundaries?

Is this still about a FFI use case? But in principal that sounds right. A library like string-parsers could internally make use of code unit level operations for performance and perhaps keep track of code unit positions but make sure code point boundaries are respected and yield safe string values.

> Why not create a separate library with safe strings and a limited API as you suggest? I suppose you would like literal syntax – I'm not sure what guarantees the PS compiler currently provides around those?

If you operate within the PureScript realm I don't think there are many reasons not to have safe strings. It is only in certain ffi situations (and performance optimizations for libraries like string-parsers) where you would need to stray from the safe path. A separate library could work I guess, but I'm not sure if it shouldn't be the other way around, a separate library for non-safe strings.

**Comment by drathier** [comment]
@chtenb 
> If the backend already exposes codepoint-level operations on the native string, why wouldn't we want to expose those through String? E.g. JS has a [codePointAt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt) function.

Because the index argument that `codePointAt` uses is a utf16 index, not a code point index. Thus, different backends will disagree. Actually fetching the n'th code point, as one might reasonably expect this function to do, is O(n) in the string length, if not cached. That's why I suggested allowing indexing into and slicing for `StringUtf16` and other encoding-specific types. The ability to O(1) index into strings is the only reason I see for wanting to use an encoding-specific type, but the performance gains can be quite dramatic, so I think it's worth it.

Re free string ffi and unpaired surrogates, I think it's fine for purescript js backend to assume all ffi strings are valid utf-16, for performance reasons. 

@MonoidMusician `CharPart` sounds exactly like a code unit, i.e. what we have today. The issue is that purescript libraries that handle code units have to assume an encoding to work with them, so naturally most choose utf16 without thinking much about it, since they're developing with the js backend, and that's the current situation we're in, where many string libraries are borkd on non-utf16 backends. The proposed solution is to be explicit about which code unit encoding you're using when you're working with code units, by using e.g. `StringUtf16` and `CharUtf16`. That way it's still free and performant on the js backend, but explicit so that non-utf16 backends can still make it work quite easily.

Re string parsers, afaict what you really want is an iterator with limited lookahead, like an iterator. One naive way would be to use code point uncons, which might generate O(n) copies of the string depending on the backend (this is what we use at work in purerl atm). Another way is to use indexes into a single immutable code unit string and make sure you read enough code units each time to get a full code point. 

Something like `getCodePointAt :: Int -> StringUtf16 -> {size :: Int, codePoint :: Char}` for O(1) indexing via code units, and returning how many code units you advanced when reading one code point, would solve that use-case fully imo. That particular function could map almost directly to `codePointAt` in js. Extending it to `getGraphemeAt :: Int -> StringUtf16 -> {size :: Int, grapheme :: Grapheme}` would also be quite easy, but that's out of scope for this change.

**Comment by chtenb** [comment]
> > > String slicing and indexing might be slow on code points based strings; implementing indexing and slicing in ffi might save some performance on utf32 backends, but probably not for js/utf16, so I'd suggest not having slicing or indexing functions on code point based strings.
> > 
> > If the backend already exposes codepoint-level operations on the native string, why wouldn't we want to expose those through String? E.g. JS has a [codePointAt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt) function.
> 
> Because the index argument that codePointAt uses is a utf16 index, not a code point index.

You're right, that's important. So my suggestion doesn't make sense.

> Re string parsers, afaict what you really want is an iterator with limited lookahead, like an iterator. One naive way would be to use code point uncons, 

Yes, that is exactly how I implemented string-parsers in the latest version to fix the `O(n^2)` performance of the code point parsers.

> which might generate O(n) copies of the string depending on the backend (this is what we use at work in purerl atm). 

I got the feeling from this thread https://github.com/purescript/purescript-strings/issues/120 that PureScript assumes the string implementation to be based on immutable strings and work with slices.

> The ability to O(1) index into strings is the only reason I see for wanting to use an encoding-specific type, but the performance gains can be quite dramatic, so I think it's worth it.

I would like to talk about the different use cases for using indexes in strings for a second.

In general, string values are black boxes until we look into them by means of a pattern search or something. We don't normally have indices pointing into the string on beforehand, we have to get them from somewhere.

If we're in the PureScript realm we could abstract over these indices by some opaque Slice object. So executing a `find :: String -> Pattern -> Slice` will give us a position into the string (with appropriate functions to leverage it), but it will not leak implementation details (along with possible confusions) about what kind of index it is. Maybe modelling such a `Slice` by a plain string is already enough (because strings are assumed to be implemented as slices), but maybe we want to store some additional information along, such as a pointer to the start of the original string, and code unit indices for the start and end of the slice which may help in defining some string library functions. The key difference between a `Slice` and an `Int` is that a slice is specific to a single string on which it operates, such that its internal indices and string slice are guaranteed to be correctly pointing to a proper sequence of code points.

If we're using a FFI and we can't help but get handed a code unit index into a string, then we may still expose appropriate functions to deal with that, like `codePointAtCodeUnitIndex :: String -> Int -> Maybe CodePoint` and `unsafeCodePointAtCodeUnitIndex :: String -> Int -> CodePoint`. These are more clunky to use (at least the safe versions because of the Maybes involved), but the idea is that their use is limited to cases where we need to bridge a gap between PureScript and FFI.

Solving both these cases satisfyingly with a single paradigm (i.e. using `Int` code unit indices everywhere) doesn't seem possible to me. Either we lose safety when indexing, or we lose performance.

**Comment by drathier** [comment]
As @chtenb points out, it would be nice to not expose encoding, and to only allow valid string offsets by construction. Let's try to reach something we can implement and ship.

I believe the two use-cases are for FFI and for parsers and similar libraries (like the Data.String.CodePoints module), and that all other users should prefer working with code points or graphemes. Please challenge me on this.

Here's an example of a bidirectional constant time iterator-based design where you cannot get half a code point (unless you use the offsets with another string):
```
null :: String -> Boolean
zeroOffset :: Offset
lastCodePointInStringOffset :: String -> Offset
codePointAt :: String -> Offset -> Char
codePointAfter :: String -> Offset -> Maybe Offset
codePointBefore :: String -> Offset -> Maybe Offset
slice :: String -> Offset -> Offset -> String
Eq Offset, Ord Offset
newtype Offset = Offset Int (private constructor)
```
This way you couldn't write code that's utf8/16/32-specific, but you still have random back-access. You have to iterate your way to the offset, but after that you have unlimited constant-time random access.
We aren't using much abstraction, so performance should be good. It's reasonable to believe that the compiler could inline these operations, like it does with many builtins.

We've also made an api that devs are unlikely to want to use, so it's unlikely to be used by accident.
If we're getting offsets from FFI, we could accept Offsets directly. It would open up the possibility of writing backend-specific code, but I guess that's what you want when you're using FFI.

Char vs CodePoint: rename CodePoint to Char, and `type CodePoint = Char` to ease the transition. Data.Char only contains functions that would work equally well on CodePoints.
Slice: I'd vote for using code point based slicing, except for FFI.
Regex: We could return Offsets from Data.String.Regex.search/searchAll and the likes.

**Comment by chtenb** [comment]
Would an Offset semantically be a Slice of length 1? I guess that makes sense, but then you can't have `zeroOffset :: String -> Offset` if the string is empty. Or if an Offset has length 0 that makes sense too, but then you can't have `codePointAt :: Offset -> Char`. I suspect an Offset definition with length 1 would be the easiest to work with, because this effectively allows you to point to the individual "units" (code points) of the string without ambiguity, but I may be wrong.

> ... (unless you use the offsets with another string):

I'm thinking out loud here, but if we make the `Offset` also reference the `String` it belongs to we can solve that for the most part. 
Then your proposed API would look like

```hs
null :: String -> Boolean
zeroOffset :: String -> Maybe Offset        -- Nothing if the string is empty
offsetAtCodePointIndex :: Int -> String -> Maybe Offset -- O(n) operation, Nothing if index out of range
lastCodePointInStringOffset :: String -> Maybe Offset -- This could be a O(1) operation, returns Nothing if string is empty
codePointAt :: Offset -> Char
offsetAfter :: Offset -> Maybe Offset
offsetBefore :: Offset -> Maybe Offset
slice :: Offset -> Offset -> Maybe String          -- Nothing if the Offsets refer to different strings
Eq Offset, Ord Offset
newtype Offset = Offset {codePointIndex :: Int, codeUnitIndex :: Int, string :: String} (private constructor)
```

But the problem of mismatched String objects remains when you have to pass multiple Offsets into a function, since there is no guarantee they refer to the same String. 

Another problem is that an Offset may not be expressive enough to be used as return value for some string search functions. Search results may be of arbitrary length (including length 0) and an Offset is defined to have length 1.

We could try to solve these issues with a `Slice` type.

```hs
atOffset :: Offset -> Slice
afterOffset :: Offset -> Slice
beforeOffset :: Offset -> Slice
fullSlice :: String -> Slice
subString :: Slice -> String           -- The string constructed by the bounds of the slice
sliceAtCodePointIndices :: Int -> Int -> String -> Slice         -- O(n) operation
length :: Slice -> Int                         -- In codepoints
find :: String -> Pattern -> Slice
startOffset :: Slice -> Offset
endOffset :: Slice -> Offset
Eq Slice, Ord Slice
newtype Slice = Slice Int Int Int Int String (private constructor)
```

However, the problem of mismatched String objects still remains when you want to do `O(1)` combinatory arithmetic with these objects. I'm not sure how many of these operations are useful. I can think of

```hs
slice :: Offset -> Offset -> Maybe Slice          -- Nothing if the Offsets refer to different strings
```

I'm not sure how much of an annoyance that `Maybe` return type is in practice. In theory we could use the String reference of the first operand and use the codepoint indices of the second to do the math. But that results in `O(n)` running times. And we probably would not want a valid return value if the offsets have mismatched strings.

```hs
slice :: Offset -> Offset -> Slice
slice offset1 offset2 = sliceAtCodePointIndices offset1.codePointIndex offset2.codePointIndex offset1.string
```

**Comment by chtenb** [comment]
> like returning Maybe's which will only be Nothing if you used it wrongly

Yeah I agree that if there is no reasonable way to handle a Nothing, the function should not return a Maybe. It's better to just leave it be a partial function.

> No, it's half a slice, or an offset/index into a string. Two offsets can form a slice [from,to].

Ok

> so I went with the simplest and presumably fastest api for it.

It could be beneficial to also track the code point offset at the cost of some overhead. If a codepoint offset needs to be reported then it doesnt have to be recomputed in O(n). But I'm not sure how beneficial this would be in practice.

---

So far we have (something along the lines of)

```hs
-- An offset refers to a position in between characters
--  h e l l o    This is a string consisting of 5 characters
-- ^ ^ ^ ^ ^ ^   These are the 6 offsets, numbered 0 to 5
newtype Offset = Offset { cuOffset :: Int, cpOffset :: Int }
startOffset :: Offset
endOffset :: String -> Offset
nextOffset :: Partial => String -> Offset -> Maybe Offset
prevOffset :: Partial => String -> Offset -> Maybe Offset
charAfter :: Partial => String -> Offset -> Maybe Char
subString :: Partial => String -> Offset -> Offset -> String
```

I've marked the functions which can exhibit undefined behavior (codepoint splicing, index out of bounds, etc) as Partial. If we were to also track the string in the Offset we would be able to get to the following API, with only the `subString` function left as Partial and with less passing around of the string. Another potential advantage would be that the undefined behavior of the Partial functions could be limited to exception throwing, since now we are able to defensively check whether the tracked strings of different offsets are the same.

```hs
newtype Offset = Offset { cuOffset :: Int, cpOffset :: Int, str :: String }
unOffset (Offset o) = o
instance Eq Offset

startOffset :: String -> Offset
endOffset :: String -> Offset
nextOffset :: Offset -> Maybe Offset
prevOffset :: Offset -> Maybe Offset
charAfter :: Offset -> Maybe Char
subString :: Partial => Offset -> Offset -> String
```

These are some edges cases and properties that I think should hold to make all the math add up.

```hs
prop_emptyStringOffset = startOffset "" == endOffset ""
prop_charAfterEmptyString = charAfter (startOffset "") == Nothing
prop_charAfterEndOffset s = charAfter (endOffset s) == Nothing

prop_prevStartOffset s = prevOffset (startOffset s) == Nothing
prop_nextEndOffset s = nextOffset (endOffset s) == Nothing
prop_iterOffset s = applyN (length s) (bindFlipped nextOffset) (Just (startOffset s)) == Just (endOffset s)

prop_emptySubString s = unsafePartial subString (startOffset s) (startOffset s) == ""
prop_negativeSubString s = unsafePartial subString (endOffset s) (startOffset s) == ""
prop_fullSubString s = unsafePartial subString (startOffset s) (endOffset s) == s
```


---

Now having an Offset value doesn't mean there are actually characters to point to, as is reflected by `prop_charAfterEndOffset`. Perhaps there are situations where it would be slightly more ergonomical to be able to actually point to a Char (codepoint) into a String and have the guarantee that it exists.

```hs
-- An index refers to a character position
--  h e l l o   This is a string consisting of 5 characters
--  ^ ^ ^ ^ ^   These are the 5 indices, numbered 0 to 4
newtype Index = Index { cuIndex :: Int, cpIndex :: Int, str :: String }
unIndex (Index o) = o

instance Eq Index

startIndex :: String -> Maybe Index
endIndex :: String -> Maybe Index
nextIndex :: Index -> Maybe Index
prevIndex :: Index -> Maybe Index
charAt :: Index -> Char

prop_emptyStringIndex = startIndex "" == Nothing && endIndex "" == Nothing
prop_prevStartIndex s = map prevIndex (startIndex s) == Nothing
prop_nextEndIndex s = map nextIndex (endIndex s) == Nothing
prop_iterIndex s = applyN (length s - 1) (bindFlipped nextIndex) (startIndex s) == endIndex s
```

I think this `Offset`/`Index` terminology is useful to distinghuish behavioral expectations. They are complementary; everything one can do with `Index` can also be done with `Offset`, so I don't think the first version of the string API needs to include the `Index` type, if at all.

The open question is: how much overhead should the Offset type have? I can't really answer that because I don't have much experience with PureScript performance. In theory it shouldn't matter, because it's a trivially small object, but in practice (perhaps even depending on the backend) there may be a lot of boxing/unboxing going. An `Int` newtype is obviously slightly leaner, but it does give in on some ergomics and safety.

**Comment by drathier** [comment]
Do we want to support lone surrogates? It could be UB, or checked in FFI or at FFI boundary. Or supported as values but not by string functions.

Let's explore the 80/20 route, what will give us the most bang for the minimal upgrade effort?

UTF-16/UCS2 (js backend) is assumed in core places in the purescript ecosystem, in purescript code, where encoding should never be assumed:
- https://github.com/purescript-contrib/purescript-string-parsers/blob/518038cec5e76a1509bab87685e0dae77462d9e1/src/StringParser/CodePoints.purs#L74
- https://github.com/purescript/purescript-strings/blob/v6.0.0/src/Data/String/CodePoints.purs#L191-L195
Fixing just these two places will get us most of the way there in supporting other encodings/backends.

Minimal api:
```
length :: String -> IndexCU
codePointAtIndex :: IndexCU -> String -> Char
codePointIndexBefore :: String -> IndexCU -> Maybe IndexCU
codePointIndexAfter :: String -> IndexCU -> Maybe IndexCU
slice :: IndexCU -> IndexCU -> String -> String

type IndexCU = Int
```
- all functions are constant time on all unicode backends
- slicing is still possible
- existing code using `Int` indices and dangerous operations like calculating indexes can continue doing so
- type signature still shows that there's something special about these `Int`s, so someone looking at the docs will know something's up
- we won't accidentally break any existing usage, since the `Int` backdoor is still there

I'm not a fan of having a function `CodePoint -> Array CodeUnit` or similar; I'd much rather just have `CodePoint -> String` or `Array CodePoint -> String` or `append :: String -> CodePoint -> String` or similar. Otherwise we're needlessly exposing the encoding again.

**Comment by MonoidMusician** [comment]
I'm not sure what you mean by this:
> we won't accidentally break any existing usage, since the `Int` backdoor is still there

The existing CodePoint API uses ints to refer to CodePoint-indices, not CodeUnit-indices. I think it would be a bit weird to mix the two, but maybe justifiable for performance. It definitely would require handling the encoding manually in the FFI though. It seems that any performant approach to CodePoints would …

Maybe it would be good just to start by documenting all the places that UTF-16 encoding is assumed.

**Comment by drathier** [comment]
@MonoidMusician 

> we won't accidentally break any existing usage, since the Int backdoor is still there

There's always a way out, i.e. some way to make any existing code work after this change.

> It seems that any performant approach to CodePoints would

Yes, the O(n) indices that the CodePoint module uses today are safer but very slow. I'm absolutely sure we want constant time access into strings, at least for low-level things like parser implementations.

> Maybe it would be good just to start by documenting all the places that UTF-16 encoding is assumed.

The two big users of utf-16 that I know of are the parser libraries: 

- https://github.com/purescript-contrib/purescript-string-parsers/blob/518038cec5e76a1509bab87685e0dae77462d9e1/src/StringParser/CodePoints.purs#L74
- https://github.com/purescript/purescript-strings/blob/v6.0.0/src/Data/String/CodePoints.purs#L191-L195

Fixing just these two places will get us most of the way there in supporting other encodings/backends.

**Comment by MonoidMusician** [comment]
Since this has been coming up again in various contexts lately, I wrote up a proposal in line with my earlier suggestions, now with details:

https://cofree.coffee/~verity/purs_unicode.html

Of note: it is not tenable to require that strings are always valid Unicode, that would require a huge reworking of the API. However, we must require that source literals are valid Unicode.

The compiler patch for this was easy. I'm starting to work through the library changes and see how much stuff will be substantially affected. I'm optimistic that upgrading will be pretty straightforward for most code.

**Comment by JordanMartinez** [comment]
> I wrote up a proposal in line with my earlier suggestions, now with details:

Any thoughts on how this would affect codecs? For example, let's say I use a typeclass-based codec to encode/decode a String. If I'm on a backend that uses UTF-8 and I send this to a backend that uses UTF-16, how will that fair?

**Comment by f-f** [comment]
> For example, let's say I use a typeclass-based codec to encode/decode a String. If I'm on a backend that uses UTF-8 and I send this to a backend that uses UTF-16, how will that fair?

Doesn't it very much depend on what "sending" means? As far as I understand here we are discussing the _internal_ representation of Strings in the runtime. Outputting data depends on how you are serialising it. If you write to a file you'll likely write it to disk as UTF-8, etc.

**Comment by MonoidMusician** [comment]
Yeah, as far as serializing goes, most input/output is in UTF-8 already (including JSON), so it doesn't support lone surrogates and nothing is changing there.

The main gotcha would be if you encode code unit _indices_ in the JSON, for some reason. That isn't going to be portable between UTF-8 and UTF-16. But there's not much we can do about that, and it seems like a pretty odd thing to do. (I certainly have not had occasion to do that.)

**Comment by JordanMartinez** [comment]
Right. Thanks all. 
Then the proposal LGTM.

**Comment by richcarl** [comment]
I looked at the proposal, but I feel that this is still too focused on individual codepoints and on indexing.

**Comment by MonoidMusician** [comment]
I don't want to have support for graphemes in the compiler and core string library.

The main reason is PureScript is still primarily a compile-to-JavaScript language, and JavaScript does not have good grapheme support. And even if it got it soon, PureScript has historically been very conservative about what JavaScript features we allow in core libraries.

I also don't think that it is a good idea for core datatypes to be tied to particular Unicode versions. (Especially since JavaScript clients can be ancient ...) This is about providing quick, well-known, and relatively safe primitives for others to be able to build upon.

A lot of programming language standards are basically specified in code points, so it is useful to have specialized literals and types for them, and fast parsers as well.

If someone wants to provide a library with graphemes, normalization functions, separate normalized string types or whatever, and wants to figure out what guarantees they want to provide around Unicode versions, that's great. It's certainly possible to implement them on top of this API, or to FFI to runtime-specific versions of them. But that's not the purpose of the core string library/type as I see it.

----

Re indexing: I'm aware that code point indexing is slow. In case it wasn't clear: in my proposed functions in the proposed `Data.String.Encoding` module, all of the indexes/lengths there are code units for efficiency. It's intended to be a useful compromise, and to work for any encoding.

The reason I still am using indexes is so that backends without fast slicing can still have performant parsing. It's easy to implement these codeunit-index-based APIs on backends with fast slicing, like JS and Erlang, but it's much harder/impossible to do the opposite (have an API that assumes fast slicing and also works well on backends without).

One specific use case is self-hosted tools like @natefaubion's [purescript-language-cst-parser](https://github.com/natefaubion/purescript-language-cst-parser) being able to work performantly on backends like the WIP Chez Scheme one, which probably will not have fast slicing (unless someone implements a new string type for it).

**Comment by drathier** [comment]
Re indices; just want to point out that all unicode encodings allow synchronization, so we can safely implement a `get the first code point on or before this index` function with almost no overhead, to avoid splitting code points using bad indices, and still get that sweet sweet O(1) index-based lookup. This one function should be enough to allow efficient parsers to be implemented in terms of code points instead of code units; get the code point on or before index 42, and tell me what size it was so I can find the index of the next code point after.

**Comment by f-f** [comment]
> Since backends like C++ already interpret Char as a full code point, why not just extend the current meaning of Char (given that we will have CodeUnit for working with individual code units), to minimize changes to existing code

The [C++ backend](https://github.com/andyarvanitis/purescript-native) is two major versions behind, and according to [the latest survey](https://raw.githubusercontent.com/purescript/survey/main/results/The%20State%20of%20PureScript%202023%20Results.pdf) a tiny amount of people (possibly one) is currently using it. The only backends that seem to get some use at this point in time are Purerl and Purenix, with Purerl2 and Purescm as upcoming ones, pending on this string rework.

All of this to say that "try to minimize code churn for the C++ backend" doesn't seem like a sensible target right now.

**Comment by richcarl** [comment]
> All of this to say that "try to minimize code churn for the C++ backend" doesn't seem like a sensible target right now.

The change to the purescript core (Char to CodePoint) would affect every piece of purs code doing anything with characters, regardless of backend.

**Comment by MonoidMusician** [comment]
re synchronization:

Yeah, I have those functions in my proposal, `syncBwd`/`syncFwd` and `advance`. They are more complicated than one would initially think because they will have to handle invalid Unicode, but I roughly specced them out.

re char:

Personally I really want to get rid of the name `Char`. There isn't a way to silently point it to `Prim.CodePoint` since it would require an import. Also because I am proposing that `Data.String.CodeUnits.CodeUnit` have a different representation, like int instead of like string, a rename of `Prim.Char` to `Data.String.CodeUnits.CodeUnit` is a good way to make that change visible, in case someone's depending on that representation. (I hope not? I can't really think of why anyone would.) It would require an import in any case.

(I'm open to keeping the representation as a string, or unspecified, if anyone feels strongly about that, but I'd still like to rename to `CodeUnit`. `CodePoint` needs to be an integer due to how codegen works.)

I think that if you were depending on char literals, the rename is incentive to update to code points instead. Having code point literals will increase the correctness and portability of most code that previously used char literals. It certainly will make code nicer that used unsafe functions to work around the lack of codepoint literals.

If you're focused on interop with JavaScript or another specific runtime, then it makes sense to keep using code units, but it's no longer the compiler's responsibility to help you write code unit literals since they are not portable.

The good news is that besides the type renames, a couple function renames, and the change in literals, most of the public string API is staying the same. That's why I wrote up my proposal. I don't expect most upgrades to be particularly onerous for these reasons.
